certora/ArrangerConduit.spec

```diff
@@ -0,0 +1,805 @@
+
+    mathint amount = min(maxAmount, maxWithdraw(ilk, asset)); 
+    withdraw(e, ilk, asset, maxAmount);
+
+    mathint balanceOfBufferAfter = gem.balanceOf(buffer);
+    mathint balanceOfConduitAfter = gem.balanceOf(currentContract);
+    mathint withdrawableFundsAfter = withdrawableFunds(asset, ilk);
+    mathint totalWithdrawableFundsAfter = totalWithdrawableFunds(asset);
+    mathint withdrawalsAfter = withdrawals(asset, ilk);
+    mathint totalWithdrawalsAfter = totalWithdrawals(asset);
+
+    mathint withdrawableFundsOtherAfter = withdrawableFunds(otherAsset, otherIlk);
+    mathint totalWithdrawableFundsOtherAfter = totalWithdrawableFunds(otherAsset2);
+    mathint withdrawalsOtherAfter = withdrawals(otherAsset3, otherIlk2);
+    mathint totalWithdrawalsOtherAfter = totalWithdrawals(otherAsset4);
+
+    assert balanceOfBufferAfter == balanceOfBufferBefore + amount, "balance of buffer did not increase by amount";
+    assert balanceOfConduitAfter == balanceOfConduitBefore - amount, "balance of conduit did not decrease by amount";
+    assert withdrawableFundsAfter == withdrawableFundsBefore - amount, "withdrawableFunds did not decrease by amount";
+    assert totalWithdrawableFundsAfter == totalWithdrawableFundsBefore - amount, "totalWithdrawableFunds did not decrease by amount";
+    assert withdrawalsAfter == withdrawalsBefore + amount, "withdrawals did not increase by amount";
+    assert totalWithdrawalsAfter == totalWithdrawalsBefore + amount, "totalWithdrawals did not increase by amount";
+
+    assert withdrawableFundsOtherAfter == withdrawableFundsOtherBefore, "other withdrawable funds changed unexpectedly";
+    assert totalWithdrawableFundsOtherAfter == totalWithdrawableFundsOtherBefore, "other total withdrawable funds changed unexpectedly";
+    assert withdrawalsOtherAfter == withdrawalsOtherBefore, "other withdrawals changed unexpectedly";
+    assert totalWithdrawalsOtherAfter == totalWithdrawalsOtherAfter, "other total withdrawals changed unexpectedly";
+}
+
+
+// Verify revert rules on withdraw
+rule withdraw_revert(bytes32 ilk, address asset, uint256 maxAmount) {
+    env e;
+
+    require asset == gem;
+
+    address buffer = registry.buffers(ilk);
+    require buffer != currentContract;
+
+    bool canCall = roles.canCall(ilk, e.msg.sender, currentContract, to_bytes4(0xa6fb97d1)); // withdraw(bytes32,address,uint256)
+    mathint balanceOfConduit = gem.balanceOf(currentContract);
+    mathint withdrawals = withdrawals(asset, ilk);
+    mathint totalWithdrawals = totalWithdrawals(asset);
+    mathint withdrawableFunds = withdrawableFunds(asset, ilk);
+    mathint totalWithdrawableFunds = totalWithdrawableFunds(asset);
+
+    mathint amount = min(maxAmount, maxWithdraw(ilk, asset));
+    withdraw@withrevert(e, ilk, asset, maxAmount);
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = !canCall;
+    bool revert3 = withdrawableFunds < amount;
+    bool revert4 = totalWithdrawableFunds < amount;
```

Comment:
This one is fine, but we might consider make an invariant proving that `totalWithdrawableFunds` equals to the sum of  `withdrawableFunds` (or at least >= than anyone). Then if we require the invariant that revert is not actually necessary.

Reply:
The invariant is basically covered in `changeTogether`.
I don't see an advantage for requiring the invariant vs having the revert condition.